survey/src/main/java/com.quickbirdstudios.surveykit/backend/views/question_parts/LocationPickerPart.kt

```diff
@@ -0,0 +1,268 @@
+package com.quickbirdstudios.surveykit.backend.views.question_parts
+
+import android.Manifest
+import android.content.Context
+import android.content.pm.PackageManager
+import android.text.Editable
+import android.text.TextWatcher
+import android.util.AttributeSet
+import android.view.Gravity
+import android.view.MotionEvent
+import android.view.View
+import android.view.animation.BounceInterpolator
+import android.view.inputmethod.InputMethodManager
+import android.widget.*
+import androidx.core.app.ActivityCompat
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import com.google.android.gms.maps.CameraUpdateFactory
+import com.google.android.gms.maps.GoogleMap
+import com.google.android.gms.maps.GoogleMapOptions
+import com.google.android.gms.maps.MapView
+import com.google.android.gms.maps.model.LatLng
+import com.quickbirdstudios.surveykit.R
+import com.quickbirdstudios.surveykit.SurveyTheme
+import com.quickbirdstudios.surveykit.backend.address.AddressSuggestionProvider
+import com.quickbirdstudios.surveykit.backend.views.main_parts.StyleablePart
+
+class LocationPickerPart @JvmOverloads constructor(
+    context: Context,
+    private val lifecycle: Lifecycle,
+    addressProvider: AddressSuggestionProvider,
+    attrs: AttributeSet? = null
+) : LinearLayout(context, attrs), StyleablePart, LifecycleObserver {
+
+    //region Members
+    private var search: AutoCompleteTextView
+    private val mapView: MapView
+    private lateinit var map: GoogleMap
```

Comment:
Since this one gets initialised in the `init` block, you do not need to `lateinit` it, but just go with val, thus you do not perform reflection on any of the variables below, `::map.isInitialized`.

Reply:
`GoogleMap` instance only can get via async `MapView.getMapAsync` function. If I can reach GoogleMap instance directly, I can go with val.
Is there any proper way to implement this? If you can explain it more, I would like to implement.

Reply:
Kotlin allows you to mark the variable as a `val` and not give a value directly but in the `init` block instead.

```kotlin
private val myObject : MyObject

init{
  myObject = giveValueToObject()
}
```

Reply:
If an instance taken by asynchronously, how can i set `val` variable in init method? Kotlin don't allows this. There is an only way to take `GoogleMap` instance like below. When I change the variable identifier `lateinit var` to `val` Kotlin can`t build.

```kotlin
mapView.getMapAsync { map ->
     this.map = map
}
```

Reply:
I also couldn't make it work, the way @coroutineDispatcher suggested. Let's leave it like that 👍